README.md
# eternal-summer-vacation

```sh
npm install eternal-summer-vacation
```

In Japan, some people stray off into August 32nd. This is believed to be the result of thaumatologic act taken by students who make resistance against arrival of September 1st, which is the date of the end of their summer vacation.

To help victims of the incident, eternal-summer-vacation provides a [Temporal](https://tc39.es/proposal-temporal/docs/) calendar that supports August 32nd and following dates. Instead of transitioning to September and so on, eternal-summer-vacation suports dates up to August 153rd.

## Usage

```ts
import { Temporal } from "@js-temporal/polyfill";
import { eternalSummerVacationCalendar } from "eternal-summer-vacation";

const today = Temporal.PlainDate.from("2021-08-31").withCalendar(
  eternalSummerVacationCalendar
);

const tomorrow = today.plus(1);

console.log(tomorrow.month, tomorrow.date); // 8, 32   ← Hooray!
```

## Example

Simple Endless Summer Vacation Calendar made with React

https://codesandbox.io/s/eternal-summer-vacation-qctks?file=/src/App.tsx

## License

MIT

## Contributing

Welcome


CHANGELOG.md
# 2.0.0 (2022-03-26)

- Upgrade `@js-temporal/polyfill` from 0.2.0 to 0.4.0

# 1.0.0

Initial release


src/usecase.test.ts
import { Temporal } from "@js-temporal/polyfill";
import { eternalSummerVacationCalendar } from "./eternal-calendar";

it("Days in 2021", () => {
  let result = "";
  let today = Temporal.PlainDate.from("2021-01-01").withCalendar(
    eternalSummerVacationCalendar
  );
  while (today.year === 2021) {
    result += `${today.year}年${today.month}月${today.day}日\n`;
    today = today.add({ days: 1 });
  }
  expect(result).toMatchInlineSnapshot(`
"2021年1月1日
2021年1月2日
2021年1月3日
2021年1月4日
2021年1月5日
2021年1月6日
2021年1月7日
2021年1月8日
2021年1月9日
2021年1月10日
2021年1月11日
2021年1月12日
2021年1月13日
2021年1月14日
2021年1月15日
2021年1月16日
2021年1月17日
2021年1月18日
2021年1月19日
2021年1月20日
2021年1月21日
2021年1月22日
2021年1月23日
2021年1月24日
2021年1月25日
2021年1月26日
2021年1月27日
2021年1月28日
2021年1月29日
2021年1月30日
2021年1月31日
2021年2月1日
2021年2月2日
2021年2月3日
2021年2月4日
2021年2月5日
2021年2月6日
2021年2月7日
2021年2月8日
2021年2月9日
2021年2月10日
2021年2月11日
2021年2月12日
2021年2月13日
2021年2月14日
2021年2月15日
2021年2月16日
2021年2月17日
2021年2月18日
2021年2月19日
2021年2月20日
2021年2月21日
2021年2月22日
2021年2月23日
2021年2月24日
2021年2月25日
2021年2月26日
2021年2月27日
2021年2月28日
2021年3月1日
2021年3月2日
2021年3月3日
2021年3月4日
2021年3月5日
2021年3月6日
2021年3月7日
2021年3月8日
2021年3月9日
2021年3月10日
2021年3月11日
2021年3月12日
2021年3月13日
2021年3月14日
2021年3月15日
2021年3月16日
2021年3月17日
2021年3月18日
2021年3月19日
2021年3月20日
2021年3月21日
2021年3月22日
2021年3月23日
2021年3月24日
2021年3月25日
2021年3月26日
2021年3月27日
2021年3月28日
2021年3月29日
2021年3月30日
2021年3月31日
2021年4月1日
2021年4月2日
2021年4月3日
2021年4月4日
2021年4月5日
2021年4月6日
2021年4月7日
2021年4月8日
2021年4月9日
2021年4月10日
2021年4月11日
2021年4月12日
2021年4月13日
2021年4月14日
2021年4月15日
2021年4月16日
2021年4月17日
2021年4月18日
2021年4月19日
2021年4月20日
2021年4月21日
2021年4月22日
2021年4月23日
2021年4月24日
2021年4月25日
2021年4月26日
2021年4月27日
2021年4月28日
2021年4月29日
2021年4月30日
2021年5月1日
2021年5月2日
2021年5月3日
2021年5月4日
2021年5月5日
2021年5月6日
2021年5月7日
2021年5月8日
2021年5月9日
2021年5月10日
2021年5月11日
2021年5月12日
2021年5月13日
2021年5月14日
2021年5月15日
2021年5月16日
2021年5月17日
2021年5月18日
2021年5月19日
2021年5月20日
2021年5月21日
2021年5月22日
2021年5月23日
2021年5月24日
2021年5月25日
2021年5月26日
2021年5月27日
2021年5月28日
2021年5月29日
2021年5月30日
2021年5月31日
2021年6月1日
2021年6月2日
2021年6月3日
2021年6月4日
2021年6月5日
2021年6月6日
2021年6月7日
2021年6月8日
2021年6月9日
2021年6月10日
2021年6月11日
2021年6月12日
2021年6月13日
2021年6月14日
2021年6月15日
2021年6月16日
2021年6月17日
2021年6月18日
2021年6月19日
2021年6月20日
2021年6月21日
2021年6月22日
2021年6月23日
2021年6月24日
2021年6月25日
2021年6月26日
2021年6月27日
2021年6月28日
2021年6月29日
2021年6月30日
2021年7月1日
2021年7月2日
2021年7月3日
2021年7月4日
2021年7月5日
2021年7月6日
2021年7月7日
2021年7月8日
2021年7月9日
2021年7月10日
2021年7月11日
2021年7月12日
2021年7月13日
2021年7月14日
2021年7月15日
2021年7月16日
2021年7月17日
2021年7月18日
2021年7月19日
2021年7月20日
2021年7月21日
2021年7月22日
2021年7月23日
2021年7月24日
2021年7月25日
2021年7月26日
2021年7月27日
2021年7月28日
2021年7月29日
2021年7月30日
2021年7月31日
2021年8月1日
2021年8月2日
2021年8月3日
2021年8月4日
2021年8月5日
2021年8月6日
2021年8月7日
2021年8月8日
2021年8月9日
2021年8月10日
2021年8月11日
2021年8月12日
2021年8月13日
2021年8月14日
2021年8月15日
2021年8月16日
2021年8月17日
2021年8月18日
2021年8月19日
2021年8月20日
2021年8月21日
2021年8月22日
2021年8月23日
2021年8月24日
2021年8月25日
2021年8月26日
2021年8月27日
2021年8月28日
2021年8月29日
2021年8月30日
2021年8月31日
2021年8月32日
2021年8月33日
2021年8月34日
2021年8月35日
2021年8月36日
2021年8月37日
2021年8月38日
2021年8月39日
2021年8月40日
2021年8月41日
2021年8月42日
2021年8月43日
2021年8月44日
2021年8月45日
2021年8月46日
2021年8月47日
2021年8月48日
2021年8月49日
2021年8月50日
2021年8月51日
2021年8月52日
2021年8月53日
2021年8月54日
2021年8月55日
2021年8月56日
2021年8月57日
2021年8月58日
2021年8月59日
2021年8月60日
2021年8月61日
2021年8月62日
2021年8月63日
2021年8月64日
2021年8月65日
2021年8月66日
2021年8月67日
2021年8月68日
2021年8月69日
2021年8月70日
2021年8月71日
2021年8月72日
2021年8月73日
2021年8月74日
2021年8月75日
2021年8月76日
2021年8月77日
2021年8月78日
2021年8月79日
2021年8月80日
2021年8月81日
2021年8月82日
2021年8月83日
2021年8月84日
2021年8月85日
2021年8月86日
2021年8月87日
2021年8月88日
2021年8月89日
2021年8月90日
2021年8月91日
2021年8月92日
2021年8月93日
2021年8月94日
2021年8月95日
2021年8月96日
2021年8月97日
2021年8月98日
2021年8月99日
2021年8月100日
2021年8月101日
2021年8月102日
2021年8月103日
2021年8月104日
2021年8月105日
2021年8月106日
2021年8月107日
2021年8月108日
2021年8月109日
2021年8月110日
2021年8月111日
2021年8月112日
2021年8月113日
2021年8月114日
2021年8月115日
2021年8月116日
2021年8月117日
2021年8月118日
2021年8月119日
2021年8月120日
2021年8月121日
2021年8月122日
2021年8月123日
2021年8月124日
2021年8月125日
2021年8月126日
2021年8月127日
2021年8月128日
2021年8月129日
2021年8月130日
2021年8月131日
2021年8月132日
2021年8月133日
2021年8月134日
2021年8月135日
2021年8月136日
2021年8月137日
2021年8月138日
2021年8月139日
2021年8月140日
2021年8月141日
2021年8月142日
2021年8月143日
2021年8月144日
2021年8月145日
2021年8月146日
2021年8月147日
2021年8月148日
2021年8月149日
2021年8月150日
2021年8月151日
2021年8月152日
2021年8月153日
"
`);
});


src/eternal-calendar.test.ts
import { eternalSummerVacationCalendar } from "./eternal-calendar";
import { Temporal } from "@js-temporal/polyfill";

const calendar = eternalSummerVacationCalendar;

describe("PlainDate withCalendar", () => {
  it("2021-01-01", () => {
    const date = Temporal.PlainDate.from("2021-01-01").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(1);
    expect(date.day).toBe(1);
  });
  it("2021-09-01", () => {
    const date = Temporal.PlainDate.from("2021-09-01").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(32);
  });
  it("2021-11-07", () => {
    const date = Temporal.PlainDate.from("2021-11-07").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(99);
  });
  it("2021-11-08", () => {
    const date = Temporal.PlainDate.from("2021-11-08").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(100);
  });
  it("2021-12-31", () => {
    const date = Temporal.PlainDate.from("2021-12-31").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(153);
  });
  it("2020-02-29", () => {
    const date = Temporal.PlainDate.from("2020-02-29").withCalendar(calendar);
    expect(date.year).toBe(2020);
    expect(date.month).toBe(2);
    expect(date.day).toBe(29);
  });
});

describe("PlainYearMonth withCalendar", () => {
  it("2021-01", () => {
    const yearMonth = Temporal.PlainYearMonth.from({
      year: 2021,
      month: 1,
      calendar,
    });
    expect(yearMonth.year).toBe(2021);
    expect(yearMonth.month).toBe(1);
  });
  it("2021-08", () => {
    const yearMonth = Temporal.PlainYearMonth.from({
      year: 2021,
      month: 8,
      calendar,
    });
    expect(yearMonth.year).toBe(2021);
    expect(yearMonth.month).toBe(8);
  });
  it("error: 2021-09", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 9,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
  it("error: 2021-10", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 10,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
  it("error: 2021-11", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 11,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
  it("error: 2021-12", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 12,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
});

describe("dateFromFields", () => {
  it("2021-01-01", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 1,
            day: 1,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-01-01[u-ca=eternal-summer-vacation]");
  });
  it("2021-08-32", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 32,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-09-01[u-ca=eternal-summer-vacation]");
  });
  it("2021-08-99", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 99,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-11-07[u-ca=eternal-summer-vacation]");
  });
  it("2021-08-100", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 100,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-11-08[u-ca=eternal-summer-vacation]");
  });
  it("2021-08-153", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 153,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-12-31[u-ca=eternal-summer-vacation]");
  });
  it("2020-02-29", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2020,
            month: 2,
            day: 29,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2020-02-29[u-ca=eternal-summer-vacation]");
  });
  it("error: 2021-09-01", () => {
    expect(() =>
      calendar.dateFromFields(
        {
          year: 2021,
          month: 9,
          day: 1,
        },
        { overflow: "constrain" }
      )
    ).toThrowError("Invalid Date");
  });
  it("error: 2021-11-06", () => {
    expect(() =>
      calendar.dateFromFields(
        {
          year: 2021,
          month: 11,
          day: 6,
        },
        { overflow: "reject" }
      )
    ).toThrowError("Invalid Date");
  });
  it("error: 2021-08-154", () => {
    expect(() =>
      calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 154,
        },
        { overflow: "reject" }
      )
    ).toThrowError("Invalid Date");
  });
});

describe("arithmetic", () => {
  describe("add: days", () => {
    it("Next day of 2021-08-31 is 2021-08-32", () => {
      const day = Temporal.PlainDate.from("2021-08-31").withCalendar(calendar);
      const tomorrow = day.add({ days: 1 });
      expect(tomorrow.month).toBe(8);
      expect(tomorrow.day).toBe(32);
    });
    it("Next day of 2021-08-99 is 2021-08-100", () => {
      const day = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 99,
        },
        { overflow: "reject" }
      );
      const tomorrow = day.add({ days: 1 });
      expect(tomorrow.month).toBe(8);
      expect(tomorrow.day).toBe(100);
    });
    it("Next day of 2021-08-153 is 2022-01-01", () => {
      const day = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 153,
        },
        { overflow: "reject" }
      );
      const tomorrow = day.add({ days: 1 });
      expect(tomorrow.year).toBe(2022);
      expect(tomorrow.month).toBe(1);
      expect(tomorrow.day).toBe(1);
    });
  });
  describe("add: months", () => {
    it("2021-01 + 1 month", () => {
      const month = new Temporal.PlainYearMonth(2021, 1, calendar);
      const result = month.add({ months: 1 });
      expect(result.year).toBe(2021);
      expect(result.month).toBe(2);
    });
    it("2021-07 + 1 month", () => {
      const month = new Temporal.PlainYearMonth(2021, 7, calendar);
      const result = month.add({ months: 1 });
      expect(result.year).toBe(2021);
      expect(result.month).toBe(8);
    });
    it("2021-08 + 1 month", () => {
      const month = new Temporal.PlainYearMonth(2021, 8, calendar);
      const result = month.add({ months: 1 });
      expect(result.year).toBe(2022);
      expect(result.month).toBe(1);
    });
    it("2021-01 - 1 month", () => {
      const month = new Temporal.PlainYearMonth(2021, 1, calendar);
      const result = month.subtract({ months: 1 });
      expect(result.year).toBe(2020);
      expect(result.month).toBe(8);
    });
    it("2021-03-05 + 5 month", () => {
      const month = new Temporal.PlainDate(2021, 3, 5, calendar);
      const result = month.add({ months: 5 });
      expect(result.year).toBe(2021);
      expect(result.month).toBe(8);
      expect(result.day).toBe(5);
    });
    it("2021-03-05 + 15 month", () => {
      const month = new Temporal.PlainDate(2021, 3, 5, calendar);
      const result = month.add({ months: 15 });
      expect(result.year).toBe(2023);
      expect(result.month).toBe(2);
      expect(result.day).toBe(5);
    });
    it("2021-08-99 + 1 month", () => {
      const month = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 99,
        },
        { overflow: "reject" }
      );
      const result = month.add({ months: 1 });
      expect(result.year).toBe(2022);
      expect(result.month).toBe(1);
      expect(result.day).toBe(31);
    });
  });
  it("until", () => {
    const day1 = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 1,
      },
      { overflow: "reject" }
    );
    const day2 = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day1.until(day2).toJSON()).toBe("P59D");
  });
});

describe("calendar properties", () => {
  it("dayOfWeek", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.dayOfWeek).toBe(3);
  });
  it("dayOfYear", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.dayOfYear).toBe(272);
  });
  it("weekOfYear", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.weekOfYear).toBe(39);
  });
  it("daysInWeek", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.daysInWeek).toBe(7);
  });
  describe("daysInMonth", () => {
    it("July", () => {
      const july = calendar.dateFromFields(
        {
          year: 2021,
          month: 7,
          day: 1,
        },
        { overflow: "reject" }
      );
      expect(july.daysInMonth).toBe(31);
    });
    it("August", () => {
      const august = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 1,
        },
        { overflow: "reject" }
      );
      console.log(august.calendar);
      expect(august.daysInMonth).toBe(153);
    });
  });
});


src/ethernal-calendar.test.ts
import { ethernalSummerVacationCalendar } from "./ethernal-calendar";
import { Temporal } from "@js-temporal/polyfill";

const calendar = ethernalSummerVacationCalendar;

describe("PlainDate withCalendar", () => {
  it("2021-01-01", () => {
    const date = Temporal.PlainDate.from("2021-01-01").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(1);
    expect(date.day).toBe(1);
  });
  it("2021-09-01", () => {
    const date = Temporal.PlainDate.from("2021-09-01").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(32);
  });
  it("2021-11-07", () => {
    const date = Temporal.PlainDate.from("2021-11-07").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(99);
  });
  it("2021-11-08", () => {
    const date = Temporal.PlainDate.from("2021-11-08").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(100);
  });
  it("2021-12-31", () => {
    const date = Temporal.PlainDate.from("2021-12-31").withCalendar(calendar);
    expect(date.year).toBe(2021);
    expect(date.month).toBe(8);
    expect(date.day).toBe(153);
  });
  it("2020-02-29", () => {
    const date = Temporal.PlainDate.from("2020-02-29").withCalendar(calendar);
    expect(date.year).toBe(2020);
    expect(date.month).toBe(2);
    expect(date.day).toBe(29);
  });
});

describe("PlainYearMonth withCalendar", () => {
  it("2021-01", () => {
    const yearMonth = Temporal.PlainYearMonth.from({
      year: 2021,
      month: 1,
      calendar,
    });
    expect(yearMonth.year).toBe(2021);
    expect(yearMonth.month).toBe(1);
  });
  it("2021-08", () => {
    const yearMonth = Temporal.PlainYearMonth.from({
      year: 2021,
      month: 8,
      calendar,
    });
    expect(yearMonth.year).toBe(2021);
    expect(yearMonth.month).toBe(8);
  });
  it("error: 2021-09", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 9,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
  it("error: 2021-10", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 10,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
  it("error: 2021-11", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 11,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
  it("error: 2021-12", () => {
    expect(() => {
      Temporal.PlainYearMonth.from({
        year: 2021,
        month: 12,
        calendar,
      });
    }).toThrow("Invalid Date");
  });
});

describe("dateFromFields", () => {
  it("2021-01-01", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 1,
            day: 1,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-01-01");
  });
  it("2021-08-32", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 32,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-09-01");
  });
  it("2021-08-99", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 99,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-11-07");
  });
  it("2021-08-100", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 100,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-11-08");
  });
  it("2021-08-153", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2021,
            month: 8,
            day: 153,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2021-12-31");
  });
  it("2020-02-29", () => {
    expect(
      calendar
        .dateFromFields(
          {
            year: 2020,
            month: 2,
            day: 29,
          },
          { overflow: "constrain" }
        )
        .toJSON()
    ).toEqual("2020-02-29");
  });
  it("error: 2021-09-01", () => {
    expect(() =>
      calendar.dateFromFields(
        {
          year: 2021,
          month: 9,
          day: 1,
        },
        { overflow: "constrain" }
      )
    ).toThrowError("Invalid Date");
  });
  it("error: 2021-11-06", () => {
    expect(() =>
      calendar.dateFromFields(
        {
          year: 2021,
          month: 11,
          day: 6,
        },
        { overflow: "reject" }
      )
    ).toThrowError("Invalid Date");
  });
  it("error: 2021-08-154", () => {
    expect(() =>
      calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 154,
        },
        { overflow: "reject" }
      )
    ).toThrowError("Invalid Date");
  });
});

describe("arithmetic", () => {
  describe("add", () => {
    it("Next day of 2021-08-31 is 2021-08-32", () => {
      const day = Temporal.PlainDate.from("2021-08-31").withCalendar(calendar);
      const tomorrow = day.add({ days: 1 });
      expect(tomorrow.month).toBe(8);
      expect(tomorrow.day).toBe(32);
    });
    it("Next day of 2021-08-99 is 2021-08-100", () => {
      const day = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 99,
        },
        { overflow: "reject" }
      );
      const tomorrow = day.add({ days: 1 });
      expect(tomorrow.month).toBe(8);
      expect(tomorrow.day).toBe(100);
    });
    it("Next day of 2021-08-153 is 2022-01-01", () => {
      const day = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 153,
        },
        { overflow: "reject" }
      );
      const tomorrow = day.add({ days: 1 });
      expect(tomorrow.year).toBe(2022);
      expect(tomorrow.month).toBe(1);
      expect(tomorrow.day).toBe(1);
    });
  });
  it("until", () => {
    const day1 = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 1,
      },
      { overflow: "reject" }
    );
    const day2 = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day1.until(day2).toJSON()).toBe("P59D");
  });
});

describe("calendar properties", () => {
  it("dayOfWeek", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.dayOfWeek).toBe(3);
  });
  it("dayOfYear", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.dayOfYear).toBe(272);
  });
  it("weekOfYear", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.weekOfYear).toBe(39);
  });
  it("daysInWeek", () => {
    const day = calendar.dateFromFields(
      {
        year: 2021,
        month: 8,
        day: 60,
      },
      { overflow: "reject" }
    );
    expect(day.daysInWeek).toBe(7);
  });
  describe("daysInMonth", () => {
    it("July", () => {
      const july = calendar.dateFromFields(
        {
          year: 2021,
          month: 7,
          day: 1,
        },
        { overflow: "reject" }
      );
      expect(july.daysInMonth).toBe(31);
    });
    it("August", () => {
      const august = calendar.dateFromFields(
        {
          year: 2021,
          month: 8,
          day: 1,
        },
        { overflow: "reject" }
      );
      console.log(august.calendar);
      expect(august.daysInMonth).toBe(153);
    });
  });
});


src/eternal-calendar.ts
import { Temporal } from "@js-temporal/polyfill";
const calendarId = "eternal-summer-vacation";
export const eternalSummerVacationCalendar =
  new (class EternalSummerVacationCalendar extends Temporal.Calendar {
    constructor() {
      super("iso8601");
    }
    override toString() {
      return calendarId;
    }
    override month(date: DateLike) {
      return eternalDate(getPlainDateWithISOCalendar(date))[0];
    }
    override day(date: DateLike) {
      return eternalDate(getPlainDateWithISOCalendar(date))[1];
    }
    override monthCode(date: DateLike) {
      const [month, , isoDate] = eternalDate(getPlainDateWithISOCalendar(date));
      return Temporal.PlainYearMonth.from({ year: isoDate.year, month })
        .monthCode;
    }
    override dateFromFields(
      fields: Parameters<typeof isoDateFromFields>[0],
      options: Temporal.AssignmentOptions
    ): Temporal.PlainDate {
      const { year, month, day } = isoDateFromFields(fields, options);
      return new Temporal.PlainDate(
        year,
        month,
        day,
        eternalSummerVacationCalendar
      );
    }
    override yearMonthFromFields(
      fields: Temporal.YearOrEraAndEraYear & Temporal.MonthOrMonthCode,
      options: Temporal.AssignmentOptions
    ): Temporal.PlainYearMonth {
      const { year, month } = isoDateFromFields({ ...fields, day: 1 }, options);
      return new Temporal.PlainYearMonth(
        year,
        month,
        eternalSummerVacationCalendar
      );
    }
    override monthDayFromFields(
      fields: Temporal.MonthOrMonthCode & {
        day: number;
      },
      options: Temporal.AssignmentOptions
    ): Temporal.PlainMonthDay {
      const { month, day } = isoDateFromFields({ ...fields, year: 0 }, options);
      return new Temporal.PlainMonthDay(
        month,
        day,
        eternalSummerVacationCalendar
      );
    }
    override daysInMonth(date: DateLike) {
      const [m, , isoDate] = eternalDate(getPlainDateWithISOCalendar(date));
      if (m === 8) {
        return AugustDays;
      }
      return super.daysInMonth(isoDate);
    }

    override dateAdd(
      dateLike: string | Temporal.PlainDate | Temporal.PlainDateLike,
      durationLike: string | Temporal.Duration | Temporal.DurationLike,
      options: Temporal.ArithmeticOptions
    ): Temporal.PlainDate {
      const date = getPlainDateWithISOCalendar(dateLike).withCalendar(
        eternalSummerVacationCalendar
      );
      const duration = Temporal.Duration.from(durationLike);

      let newYear = date.year + duration.years;
      let newMonth = date.month + duration.months;
      // -15 -- -8 -> -2
      //  -7 --  0 -> -1
      //   1 --  8 -> 0
      //   9 -- 15 -> +1
      //  16 -- 23 -> +2 ...
      const diffy = Math.floor((newMonth - 1) / 8);
      newYear += diffy;
      newMonth -= diffy * 8;

      let newDay = date.day;
      if (duration.months) {
        // if months changed, we need to check if the day is still valid
        const yearMonth = eternalSummerVacationCalendar.yearMonthFromFields(
          {
            year: newYear,
            month: newMonth,
          },
          { overflow: "reject" }
        );
        if (options.overflow === "reject") {
          if (date.day > yearMonth.daysInMonth) {
            throw new RangeError(
              `${date.day} is not a valid day for ${yearMonth.monthCode}`
            );
          }
        } else {
          if (newDay > yearMonth.daysInMonth) {
            newDay = yearMonth.daysInMonth;
          }
        }
      }

      newDay += duration.days;
      // TODO: this loop may be heavy
      do {
        if (newDay < 1) {
          newMonth--;
          if (newMonth < 1) {
            newYear--;
            newMonth = 8;
          }
          const prevYearMonth =
            eternalSummerVacationCalendar.yearMonthFromFields(
              {
                year: newYear,
                month: newMonth,
              },
              { overflow: "reject" }
            );
          newDay = prevYearMonth.daysInMonth + newDay;
          continue;
        }
        const yearMonth = eternalSummerVacationCalendar.yearMonthFromFields(
          {
            year: newYear,
            month: newMonth,
          },
          { overflow: "reject" }
        );
        if (newDay > yearMonth.daysInMonth) {
          newDay -= yearMonth.daysInMonth;
          newMonth++;
          if (newMonth > 8) {
            newYear++;
            newMonth = 1;
          }
          continue;
        }
        break;
      } while (true);

      return Temporal.PlainDate.from({
        year: newYear,
        month: newMonth,
        day: newDay,
        calendar: eternalSummerVacationCalendar,
      });
    }

    override fields(fields: Iterable<string>) {
      return super.fields(fields);
    }
  })();

const AugustDays = 31 + 30 + 31 + 30 + 31;

type DateLike =
  | Temporal.PlainDate
  | Temporal.PlainDateTime
  | Temporal.PlainYearMonth
  | Temporal.PlainDateLike
  | string;

function eternalDate(
  isoDate: Temporal.PlainDate
): [month: number, day: number, isoDate: Temporal.PlainDate] {
  if (isoDate.month < 8) {
    return [isoDate.month, isoDate.day, isoDate];
  }
  const august1st = isoDate.with({ month: 8, day: 1 });
  const daysFromAugust1st =
    isoDate.since(august1st, { largestUnit: "day" }).days + 1;
  return [8, daysFromAugust1st, isoDate];
}

function getPlainDateWithISOCalendar(date: DateLike) {
  if (isTemporalDateLike(date)) {
    const fields = date.getISOFields();
    return new Temporal.PlainDate(
      fields.isoYear,
      fields.isoMonth,
      fields.isoDay
    );
  } else if (typeof date === "string") {
    return Temporal.PlainDate.from(date);
  } else {
    return Temporal.PlainDate.from({ ...date, calendar: "iso8601" });
  }
}

function constrain(num: number, min: number, max: number) {
  if (num < min) {
    return min;
  }
  if (num > max) {
    return max;
  }
  return num;
}

/**
 * Converts an eternal date to ISO date.
 */
function isoDateFromFields(
  fields: Temporal.YearOrEraAndEraYear &
    Temporal.MonthOrMonthCode & { day: number },
  options: Temporal.AssignmentOptions
): {
  year: number;
  month: number;
  day: number;
} {
  const month = hasNumberProperty(fields, "month")
    ? fields.month
    : monthCodeToMonth(fields.monthCode);
  const year = hasNumberProperty(fields, "year") ? fields.year : fields.eraYear;
  let day = fields.day;
  if (month === 8) {
    if ((day < 1 || AugustDays < day) && options.overflow === "reject") {
      throw new RangeError("Invalid Date");
    }
    day = constrain(day, 1, AugustDays);
    const august1st = new Temporal.PlainDate(year, 8, 1);
    const targetDate = august1st.add({ days: day - 1 });
    return {
      year: targetDate.year,
      month: targetDate.month,
      day: targetDate.day,
    };
  }
  if (month > 8) {
    throw new RangeError("Invalid Date");
  }
  return {
    year,
    month,
    day: day ?? 1,
  };
}

function isTemporalDateLike(
  date: DateLike
): date is
  | Temporal.PlainDate
  | Temporal.PlainYearMonth
  | Temporal.PlainMonthDay {
  return (
    date instanceof Temporal.PlainDate ||
    date instanceof Temporal.PlainYearMonth ||
    date instanceof Temporal.PlainMonthDay
  );
}

function isPropertyAccessible(obj: unknown): obj is Record<string, unknown> {
  return obj !== null && obj !== undefined;
}

function hasNumberProperty<K extends string>(
  obj: unknown,
  key: K
): obj is Record<K, number> {
  return isPropertyAccessible(obj) && typeof obj[key] === "number";
}

function monthCodeToMonth(monthCode: string): number {
  if (monthCode.startsWith("M")) {
    return parseInt(monthCode.slice(1), 10);
  }
  throw new Error("Invalid month code: " + monthCode);
}
